Cluster operations (#189)
iomesh/docs: docs/advanced-functions/multiple-cluster-management.md

---
id: manage-multiple-cluster
title: Multiple Cluster Management
sidebar_label: Multiple Cluster Management
---

> In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation, and each IOMesh cluster is an independent storage pool. In this case, the IOMesh CSI driver enables different IOMesh clusters to communicate, reducing the number of CSI drivers that may consume unnecessary node resources.

Author: ”链接不同改动 IOMesh 集群“是啥意思？

> In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation, and each IOMesh cluster is an independent storage pool. In this case, the IOMesh CSI driver enables different IOMesh clusters to communicate, reducing the number of CSI drivers that may consume unnecessary node resources.

Reviewer: 嗯，我也觉得这里 to communicate 不太对，是不是可以改为不同的 IOMesh 集群共用 CSI Driver？

> In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation, and each IOMesh cluster is an independent storage pool. In this case, the IOMesh CSI driver enables different IOMesh clusters to communicate, reducing the number of CSI drivers that may consume unnecessary node resources.

Reviewer: 另外，reducing the number of CSI drivers  这个我建议修改为 减少 CSI Driver 的 Controller Plugin Pod 的数量，因为下面的图其实也看不出具体是少了什么

> In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation, and each IOMesh cluster is an independent storage pool. In this case, the IOMesh CSI driver enables different IOMesh clusters to communicate, reducing the number of CSI drivers that may consume unnecessary node resources.

Author: 【减少 CSI Driver 的 Controller Plugin Pod 的数量】需要这么细吗？当时问 @gxglls 这里的 benefit是从多个 CSI 减到只用我们的一个 IOMesh CSI driver 就可以了，减少资源消耗

> In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation, and each IOMesh cluster is an independent storage pool. In this case, the IOMesh CSI driver enables different IOMesh clusters to communicate, reducing the number of CSI drivers that may consume unnecessary node resources.

Reviewer: 多个 CSI 也是我们的 CSI，这里是不是有什么搞混了

> In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation, and each IOMesh cluster is an independent storage pool. In this case, the IOMesh CSI driver enables different IOMesh clusters to communicate, reducing the number of CSI drivers that may consume unnecessary node resources.

Reviewer 2: 这里是减少  Controller Plugin Pod 的数量。没有多个 csi 的概念，只有 IOMesh CSI driver

> In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation, and each IOMesh cluster is an independent storage pool. In this case, the IOMesh CSI driver enables different IOMesh clusters to communicate, reducing the number of CSI drivers that may consume unnecessary node resources.

Author: In a large-scale Kubernetes cluster, you can deploy multiple IOMesh clusters for data isolation. In this case, there is typically one management cluster that holds components such as the IOMesh Operator, IOMesh CSI Driver, and Node Disk Manager, while other clusters function as independent storage pools without any management responsibilities. 

The IOMesh CSI driver is shared by all IOMesh clusters to facilitate connection, which reduces the number of Controller Plugin pods for the CSI driver.

![image](https://user-images.githubusercontent.com/102718816/228175494-9d69fac5-de12-4519-a85f-2520c2070f4c.png)

## Deployment

> To reduce the number of Pods required in a multi-cluster deployment of IOMesh, the management components shared by all IOMesh clusters, including the IOMesh Operator, IOMesh CSI driver, and Node Disk Manager, will be installed on the first IOMesh cluster, which is referred to as the management cluster.

Reviewer: 建议调整：https://smartx1.slack.com/archives/C04AW6VDKF1/p1680752858292449?thread_ts=1680254292.921529&cid=C04AW6VDKF1

Author: 【operator 和 ndm 需要确认下】 @gxglls 子银看看？

Reviewer 2: operator 和 ndm 是共享的

**Prerequisites**
- Verify that all requirements in [Prerequisites](../deploy-iomesh-cluster/prerequisites.md) are met.

> - The IOMesh version should be 1.0.0. 

Author: done

- A Kubernetes cluster consisting of at least 6 worker nodes.

**Procedure**

The following section assumes you have 6 worker nodes, deploying the first cluster `iomesh-cluster` on worker nodes `k8s-worker-0`, `k8s-worker-1`, and `k8s-worker-2` and the second cluster `iomesh-cluster-1` on worker nodes `k8s-worker-3`, `k8s-worker-4`, and `k8s-worker-5`. 

### Deploy First IOMesh Cluster

> 1. Set up [`open-iscsi`](../deploy-iomesh-cluster/setup-worker-node.md) on each worker node.

Author: 设置工作节点从 prerequisites 里面拆出来了

2. Export the YAML config `iomesh.yaml`. 

    ```
    helm show values iomesh/iomesh > iomesh.yaml
    ```
3. Configure `iomesh.yaml`.

    - Set the field [`iomesh.chunk.dataCIDR`](../deploy-iomesh-cluster/prerequisites.md#network-requirements) to the CIDR you configured for the IOMesh storage network.

        ```yaml
        iomesh:
            chunk:

>             dataCIDR: <your-data-cidr-here>

Author: done

>         ```

Author: @gxglls 这里部署单集群的 安装 helm、 添加 iomesh helm repo、选择 DICR、全闪混闪、iomesh 版本，是不是都需要呢？ @gxglls

   iomesh-meta-1                                         2/2     Running   0          5h35m
   iomesh-meta-2                                         2/2     Running   0          5h34m
   iomesh-openebs-ndm-cluster-exporter-68c757948-qmtt4   1/1     Running   0          31h
   iomesh-openebs-ndm-gnkzt                              1/1     Running   0          31h
   iomesh-openebs-ndm-lrww8                              1/1     Running   0          31h
   iomesh-openebs-ndm-node-exporter-rp4qr                1/1     Running   0          31h
   iomesh-openebs-ndm-node-exporter-sgzpp                1/1     Running   0          31h
   iomesh-openebs-ndm-node-exporter-txbn5                1/1     Running   0          31h
   iomesh-openebs-ndm-operator-584fdbcb94-2tjsp          1/1     Running   0          31h
   iomesh-openebs-ndm-zdngg                              1/1     Running   0          31h
   iomesh-zookeeper-0                                    1/1     Running   0          31h
   iomesh-zookeeper-operator-64957fcc4f-zpdjp            1/1     Running   0          30h
   operator-f5644b7f9-2vvw7                              1/1     Running   0          5h37m
   ```

> ### Deploy Second IOMesh Cluster

Reviewer: 如果是一开始按照 install iomesh 里部署了集群，之后想部署第二个，该怎么做

Author: 不支持从一个扩到多个，支持从多个集群扩，比如从 2 个扩到 3 个 @gxglls

Reviewer: 那在单集群的安装那里是不是加一个提示比较好？如果有部署多集群的计划，建议按照多集群部署里部署第一个集群的步骤部署第一个集群，这个建议可行吗？

Reviewer 2: 目前 “从一个扩到多个” 这个功能我想了下已经是可以支持的，我提个 pr 去掉这个限制

Author: 这个 pr 有下文没？ @gxglls

                  - matchExpressions:
                  - key: kubernetes.io/hostname # The key of the Kubernetes node label.
                      operator: In
                      values: # The values of the Kubernetes node label.
                      - k8s-woker-3
                      - k8s-woker-4
                      - k8s-woker-5
              podAntiAffinity:
              preferredDuringSchedulingIgnoredDuringExecution:
              - podAffinityTerm:
                  labelSelector:
                      matchExpressions:
                      - key: app
                      operator: In
                      values:
                      - iomesh-cluster-1-zookeeper
                  topologyKey: kubernetes.io/hostname
                  weight: 20
          securityContext:
              runAsUser: 0
          persistence:
          reclaimPolicy: Delete
          spec:
              storageClassName: hostpath
              resources:
              requests:
                  storage: 20Gi
      ```
    - Apply the YAMl config to create the `zookeeper` cluster.
      ```shell
      kubectl apply -f iomesh-cluster-1-zookeeper.yaml
      ```

> 3. Create `iomesh-cluster-1.yaml` and configure `meta.nodeAffinity`, `chunk.dataCIDR`, `chunk.blockDeviceNamespace`, `chunk.nodeAffinity`, `redirector.dataCIDR`, and `redirector.nodeAffinity`.

Reviewer 3: 应该还有 diskDeploymentMode？

    - Mandatory: Fill in `dataCIDR` for `chunk` and `redirector` with the IOMesh dataCIDR.
    - Mandatory: Set `spec.chunk.devicemanager.blockDeviceNamespace` to `iomesh-system` as management components are installed in the namespace `iomesh-system` and all block devices are also in this namespace.

>     - Optional. `iomesh-cluster-1.yaml` installs IOMesh Community Edition by default. To set it to Enterprise Edition, set `image.repository.tag` to `v5.3.0-rc13-enterprise` for `meta`, `chunk`, and `redirector` respectively.

Reviewer 3: 设置 Edition 应该算 Mandatory

                  operator: In
                  values:  # The values of the Kubernetes node label.
                        - k8s-woker-3
                        - k8s-woker-4
                        - k8s-woker-5
          probe:
            image:
              repository: iomesh/operator-probe
              tag: v1.0.0
              pullPolicy: IfNotPresent
          toolbox:
            image:
              repository: iomesh/operator-toolbox
              tag: v1.0.0
              pullPolicy: IfNotPresent

      ```
### Mount Disks

1. [View block devices available for use](../deploy-iomesh-cluster/setup-iomesh.md#view-block-device-objects). Note that all block devices resides in the namespace `iomesh-system`.

    ```shell
    kubectl --namespace iomesh-system -o wide get blockdevice
    ```

2. [Configure DeviceMap](../deploy-iomesh-cluster/setup-iomesh.md#configure-devicemap).

    ```shell
    kubectl edit iomesh -n iomesh-system # Configure deviceMap for the first IOMesh cluster.
    kubectl edit iomesh-cluster-1 -n iomesh-cluster-1 # Configure deviceMap for the second IOMesh cluster.

### Configure Multiple-Cluster Connection

> To enable the IOMesh CSI driver to connect to multiple IOMesh clusters, you need to configure a `ConfigMap` containing connection information for each IOMesh cluster. Before configuration, familarity with `ConfigMap` is suggested.

Reviewer 2: ```suggestion
To enable the IOMesh CSI driver to connect to multiple IOMesh clusters, you need to create a `ConfigMap` containing connection information for each IOMesh cluster. Before configuration, familarity with `ConfigMap` is suggested.
```

> `ConfigMap`

Author: 这里是想让用户先了解这些字段，再开始配置

| Field | Description |
| ----- | ----- |
| `data.clusterId` | The Kubernetes cluster ID, which is customizable. Since a Kubernetes cluster can only have one cluster ID, two iomesh clusters deployed in the same Kubernetes cluster must have the same field value filled in.|
| `data.iscsiPortal`| iSCSI access point, fixed to 127.0.0.1:3260.|
| `data.metaAddr`      | IOMesh meta service address, which follows the format: `<iomesh-cluster-name>-meta-client.<iomesh-cluster-namespace>.svc.cluster.local:10100.` |

**Procedure**

1. Create `ConfigMap` for the first IOMesh cluster.

    Create a YAML config `iomesh-csi-configmap.yaml` with the following content. Then apply the YAML config to generate `ConfigMap`. 

      ```yaml
      apiVersion: v1
      kind: ConfigMap
      metadata:
        name: iomesh-csi-configmap
        namespace: iomesh-system
      data:
        clusterId: k8s-cluster
        iscsiPortal: 127.0.0.1:3260

>         metaAddr: iomesh-meta-client.iomesh-system.svc.cluster.local:10100

Reviewer 3: 上面的字段表里是 iomesh-cluster-name-meta-client.iomesh-cluster-namespace.svc.cluster.local:10100.
这里给的示例里是不是没写全 iomesh-cluster-name；第一个 IOmesh 集群的名称是 **iomesh-cluster**

Reviewer 2: 第一个集群的名字是 iomesh

Reviewer 3: <img width="825" alt="image" src="https://user-images.githubusercontent.com/124547192/230277443-3bc119fe-198f-49af-9379-959f6b99c12e.png">
@gxglls @yuqiu123 那这里前后不一致了，Procedure 写的是 iomesh cluster；要不在 procedure里明确一下两个集群各自的名称和 Namespace

Author: done

      ``` 
      ```shell
      kubectl apply -f iomesh-csi-configmap.yaml

> 

Reviewer 3: ```suggestion
```

Author: done

    Create a YAML config `iomesh-cluster-1-csi-configmap.yaml` with the following content. Then apply the YAML config to generate `ConfigMap`.

      ```
      apiVersion: v1
      kind: ConfigMap
      metadata:
        name: iomesh-cluster-1-csi-configmap
        namespace: iomesh-cluster-1
      data:
        clusterId: k8s-cluster
        iscsiPortal: 127.0.0.1:3260
        metaAddr: iomesh-cluster-1-meta-client.iomesh-cluster-1.svc.cluster.local:10100
      ```
      ```shell
      kubectl apply -f iomesh-cluster-1-csi-configmap.yaml
      ```

### Create StorageClass for Each Cluster

When deploying more than one IOMesh cluster, you must create a separate StorageClass for each cluster, rather than using the default StorageClass `iomesh-csi-driver`. For details, refer to the following and [Create StorageClass](../volume-operations/create-storageclass.md). 

|Field| Description|
|---|---|
| `parameters.clusterConnection` |The namespace you specify in `configMap`/the `configMap` name.|
| `parameters.iomeshCluster`| The namespace where the IOMesh cluster is located/the IOMesh cluster name.                              |

**Procedure**

> 1. Create a StorageClass for the first IOMesh cluster.

Reviewer 3: 这里也是，第一个集群存储类是 iomesh，第二个集群存储类是 iomesh-cluster-1；是不是可以调整成一个系统下的命名方式

Author: 之前有跟 ziyin 说过这个问题， @gxglls 再看看咧

Reviewer 2: 这里是有意为之的，目前第一个集群叫 iomesh 可以大幅减少部署步骤

Author: @yixin-chen-23 改成这样了，看看是否可行？http://192.168.26.13:3000/next/advanced-functions/manage-multiple-cluster#:~:text=each%20worker%20node.-,Procedure,Deploy%20Management%20Cluster,-Export%20the%20YAML

Reviewer 3: @yuqiu123 iomesh-cluster-1 的 role 叫 independent storage cluster 或者 就 non-management cluster 呢？之前 @gxglls 子银说现在对外不暴露存储池这个概念

      ```yaml
      apiVersion: storage.k8s.io/v1
      kind: StorageClass
      metadata:
        name: iomesh
      parameters:
        csi.storage.k8s.io/fstype: "ex4" 
        replicaFactor: "2" # 
        thinProvision: "true"
        reclaimPolicy: Delete
        clusterConnection: "iomesh-system/iomesh-csi-configmap"  # Specify the namespace and configMap of the first IOMesh cluster.

>         iomeshCluster: "iomesh-system/iomesh"

Reviewer 3: ```suggestion
        iomeshCluster: "iomesh-system/iomesh-cluster"
```

    volumeBindingMode: Immediate
    provisioner: com.iomesh.csi-driver
    allowVolumeExpansion: true
    ```
      ```shell
      kubectl apply -f <yaml.filename> 
      ```

### Verify Deployment 

To verify if the IOMesh clusters are deployed, create a PVC using the StorageClass you created respectively. 

1. Create a PVC for the first IOMesh cluster.

    ```yaml
    kind: PersistentVolumeClaim
    apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: iomesh-pvc
    spec:
      storageClassName: iomesh
      accessModes:
        - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
    ```

    ```shell
    kubectl apply -f <yaml.filename>
    ```

2. Create a PVC for the second IOMesh cluster. 

      ```yaml
      kind: PersistentVolumeClaim
      apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        name: iomesh-cluster1-pvc
      spec:
        storageClassName: iomesh-cluster-1
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi
      ```                                               

      ```shell
      kubectl apply -f <yaml.filename>
      ```

> IOMesh automatically enables typology awareness to ensure correct pod scheduling. When a PVC is created in the first IOMesh cluster, the pod using it is scheduled to the worker node in the same cluster for I/O localization.

Reviewer 2: ```suggestion
IOMesh automatically enables topology awareness to ensure correct pod scheduling. When a PVC is created in the first IOMesh cluster, the pod using it is scheduled to the worker node in the same cluster for I/O localization.
```

## Operations & Management 

> All procedures below are listed based on the example in [Multiple Cluster Deployment](../advanced-functions/multiple-cluster-management.md#deployment).

Reviewer: 这里的 link 和文字要修改下？

Author: done http://192.168.26.13:3000/next/advanced-functions/manage-multiple-cluster#configure-multiple-cluster-connection

### Upgrade Multiple Clusters

> _Note:_

> > When upgrading multiple IOMesh clusters, upgrade the management cluster first and then the other clusters. If not, the non-management cluster will be temporarily unavailable during the second upgrade, but all clusters will return to normal afterwards.

Reviewer: 在 install 的时候建议就把 management cluster 和 non-management cluster 的概念介绍一下

Author: 这里说了 
![image](https://user-images.githubusercontent.com/102718816/230762008-60a57ff7-f587-4b39-90ea-6fae4362471b.png)

Reviewer: 用户在之后的运维里如何判断呢？

Author: 这里重写了一下，看看是不是好读一点？http://192.168.26.13:3000/next/advanced-functions/manage-multiple-cluster

**Procedure**

1. [Upgrade the management cluster](../cluster-operations/upgrade-cluster.md), which is the first IOMesh cluster.

2. After the first upgrade is complete, edit the second IOMesh cluster.

    - View YAML config for the first IOMesh cluster.
      ```bash

>       kubectl edit iomesh iomesh-cluster-1 -n iomesh-cluster-1

Reviewer 3: 这条命令应该错了；和下面那条一样了

      ```

    - Edit the second IOMesh cluster, ensuring all values of `spec.*.image.tags` are consistent with the values of the first cluster.
      ```
      kubectl edit iomesh iomesh-cluster-1 -n iomesh-cluster-1
      ```

### Scale Multiple Clusters

There is no difference between scaling up one cluster and scaling up multiple cluster. Plan the number of worker nodes and increase the number of meta or chunk pods one by one. For more information, see [Scale IOMesh Cluster](../cluster-operations/scale-cluster.md).

### Uninstall Multiple Clusters

> When uninstalling more than one IOMesh cluster, uninstall the other clusters first and then the management cluster last. If you do not follow the order, there may be resources resided in the namespace `iomesh-system`, which may affect next deployment in this namespace.

Reviewer: ```suggestion
When uninstalling more than one IOMesh cluster, uninstall the non-management clusters first and then the management cluster last. If you do not follow the order, there may be resources resided in the namespace `iomesh-system`, which may affect the next deployment in this namespace.
```

**Procedure**

> 1. Uninstall the second IOMesh cluster, which will also delete `iomesh` and `zookeeper` components. 

Reviewer: ```suggestion
1. Uninstall the non-management IOMesh cluster, which will also delete `iomesh` and `zookeeper` components. 
```

    To uninstall another cluster, replace `iomesh-cluster-1-zookeeper.yaml` with the zookeeper YAML filename and `iomesh-cluster-1.yaml` with the YAML config name.

    ```shell
    kubectl delete -f iomesh-cluster-1-zookeeper.yaml && kubectl delete -f iomesh-cluster-1.yaml
    ```

> 2. [Uninstall the first IOMesh Cluster](../cluster-operations/uninstall-cluster.md).

Reviewer: ```suggestion
2. [Uninstall the management IOMesh Cluster](../cluster-operations/uninstall-cluster.md).
```

    ```shell
    helm uninstall --namespace iomesh-system iomesh
    ```

### License Management

> Each IOMesh cluster has a license with a unique serial number, and update the license from `Trial` to `Subscription` or `Perpetual` for each IOMesh cluster respectively.For other operations, refer to [Manage License](../cluster-operations/manage-license.md).

Reviewer: ```suggestion
Each IOMesh cluster has a license with a unique serial number, and you need to update the license from `Trial` to `Subscription` or `Perpetual` for each IOMesh cluster respectively. For other operations, refer to [Manage License](../cluster-operations/manage-license.md).
```
